geom/twkb_parser.go

```diff
@@ -0,0 +1,528 @@
+package geom
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"math"
+)
+
+// Tiny Well Known Binary
+// See spec https://github.com/TWKB/Specification/blob/master/twkb.md
+
+// UnmarshalTWKB parses a Tiny Well Known Binary (TWKB), returning the
+// corresponding Geometry.
+func UnmarshalTWKB(twkb []byte, opts ...ConstructorOption) (Geometry, error) {
+	p := newtwkbParser(twkb, opts...)
+	g, err := p.nextGeometry()
+	return g, p.annotateError(err)
+}
+
+// UnmarshalTWKBWithHeaders parses a Tiny Well Known Binary (TWKB),
+// returning the corresponding Geometry, and any bounding box and any IDs
+// listed in its header information.
+func UnmarshalTWKBWithHeaders(twkb []byte, opts ...ConstructorOption) (g Geometry, bbox []int64, ids []int64, err error) {
```

Comment:
Your understanding of the bounding box situation is correct in all ways. I'll add the suggested cheap function for just reading the bounding box header if it exists in the data stream.

Reply:
Concerning the bounding box header parsing:

I've modified the function signature for
```UnmarshalTWKBWithHeaders(twkb []byte, opts ...ConstructorOption) (g Geometry, bbox []Point, ids []int64, err error)```

Previously that function was returning bounding box information as integers (which exposed implementation internals, naughty!) but now the bounding box is returned as a slice of `Point`. If the slice is empty, there was no bounding box header. Otherwise it contains a minimum and a maximum `Point`. Because `Point` is used, the X, Y, Z and M coordinate values may all be present; no data loss occurs.

I've also added two new functions:
```UnmarshalTWKBBoundingBoxHeader(twkb []byte) (bbox []Point, err error)```
and
```UnmarshalTWKBEnvelope(twkb []byte) (bbox Envelope, hasBBox bool, err error)```

The former function only parses the bounding box header, so it returns without parsing the full geometry.
The semantics of the `Point` slice are the same as discussed above.

The latter function also only parses the bounding box header, but returns it as an `Envelope`.
The `hasBBox` flag is false if there's no such header. Otherwise, the envelope is valid.
Due to the definition of `Envelope` only X and Y values can be reported this way.
(Any Z and M values are omitted because they cannot be represented in an `Envelope`.)

I've added tests checking all these functions behave correctly.